psl-core/src/main/java/org/linqs/psl/evaluation/statistics/RankingEvaluator.java

```diff
@@ -0,0 +1,211 @@
+    }
+
+    public RankingEvaluator(String representative, int... rawCategoryIndexes) {
+        this(RepresentativeMetric.valueOf(representative.toUpperCase()), rawCategoryIndexes);
+    }
+
+    public RankingEvaluator(RepresentativeMetric representative, int... rawCategoryIndexes) {
+        this.representative = representative;
+
+        categoryIndexes = new ArrayList<Integer>(rawCategoryIndexes.length);
+
+        for (int catIndex : rawCategoryIndexes) {
+            categoryIndexes.add(catIndex);
+        }
+
+        threshold = Options.EVAL_RANK_THRESHOLD.getDouble();
+        defaultPredicate = Options.EVAL_RANK_DEFAULT_PREDICATE.getString();
+    }
+
+    /**
+     * Grab the Constant arguments of an atom at specified positions.
+     */
+    private List<Constant> getArgsAtPositions(GroundAtom atom, List<Integer> indexes) {
+        Constant[] atomArgs = atom.getArguments();
+        List<Constant> argsAtPositions = new ArrayList<Constant>(indexes.size());
+        for (int index : indexes) {
+            argsAtPositions.add(atomArgs[index]);
+        }
+        return argsAtPositions;
+    }
+
+    @Override
+    public void compute(TrainingMap trainingMap) {
+        if (defaultPredicate == null) {
+            throw new UnsupportedOperationException("RankingEvaluators must have a default predicate set (through config).");
+        }
+
+        compute(trainingMap, StandardPredicate.get(defaultPredicate));
+    }
+
+    @Override
+    public void compute(TrainingMap trainingMap, StandardPredicate predicate) {
+        labelMap = trainingMap.getLabelMap();
+        rankMap = new HashMap<GroundAtom, Integer>();
+
+        int predicateArity = predicate.getArity();
+        List<Integer> entityIndexes = new ArrayList<Integer>(predicateArity - categoryIndexes.size());
+
+        for (int index = 0; index < predicateArity; index++) {
+            if (categoryIndexes.contains(index)) {
+                continue;
+            } else {
+                entityIndexes.add(index);
+            }
+        }
+
+        // We keep a map from entities to Lists of all GroundAtoms which contain that entity.
+        Map<List<Constant>, List<GroundAtom>> sortedAtoms = new HashMap<List<Constant>, List<GroundAtom>>();
+
+        // Partition the predicted atoms into Lists
+        // according to what entity they contain.
+        for (GroundAtom atom : trainingMap.getAllPredictions()) {
+            if (atom.getPredicate() != predicate) {
+                continue;
+            }
+
+            List<Constant> entity = getArgsAtPositions(atom, entityIndexes);
+
+            if (!sortedAtoms.containsKey(entity)) {
+                sortedAtoms.put(entity, new ArrayList<GroundAtom>());
+            }
+
+            sortedAtoms.get(entity).add(atom);
+        }
+
+        // Sort these Lists and populate the rankMap.
+        for (List<Constant> entity : sortedAtoms.keySet()) {
+            Collections.sort(sortedAtoms.get(entity));
```

Comment:
The direction of the sort matters here.
There will probably need to be a boolean config option describing which direction to sort (defaulting to high values first).